Experiment: Create timing report.

ehuss opened the pull request:
This is just an experiment, so I'm not sure if we'll want to merge it.

This adds an HTML report which gets saved to disk when the build is finished.  It is primarily geared for identifying slow dependencies, and for visualizing how pipelining affects the build.

Here's an example: https://ehuss.github.io/cargo-timing.html
You can mouse over the blocks to highlight the reverse-dependencies that are released when a unit finishes.  `syn` is a really good example.

It does a few other things, like displaying a message after each unit is finished.  See the docs for more information.

rust-highfive commented:
r? @nrc

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #7216) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
> Have you checked around to see if there's any sort of open formats for this sort of data? 

FWIW I was curious so I was poking around with this today. I was thinking of `about:tracing` and Chrome's [tracing format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#), but I got some pretty disappointing graphs out of it and it didn't really give me (by default at least) any tools to work with. I wasn't able to really learn much from this output.

I also read over [Firefox's format](https://github.com/firefox-devtools/profiler/blob/master/docs-developer/gecko-profile-format.md) and didn't dive too deep into it but figured it wouldn't work.

All in all I don't think we'll get much use out of trying to use an open format.

---

One idea I had though was inspired by [this post](https://aras-p.info/blog/2017/08/08/Unreasonable-Effectiveness-of-Profilers/) where it'd be ideal if we could flatten this whole graph to see a more linear sequence of events rather than a huge waterfall. We could ideally say what core each process started running on and associate that with rows, so everything would be a bit more compact and you might be able to more visually see idle parts of the build.

If you want to get crazy it'd be pretty sweet to overlay "what's currently building" along with CPU usage over time so we could get an idea of when the CPU is idle, what's building, and what it's bottlenecked on.

To reiterate though there's tons of directions we could take it, and landing anything is better than pontificating about how we could improve it, so I'd lean on the side of landing what we have vs wondering how to implement more fancy features.

ehuss commented:
> How come only some of the dependency edges showed up when you hover over things?

Right, for the regex-syntax case, `regex` was still waiting on `aho-corasick`.  I didn't want to show all edges because the number of lines would be overwhelming.  It's already a bit of a mess.

> Printing out a "critical path" would be pretty neat for easy diagnosis.

What does this mean?  Can you give an example of what it would say?  It seems like there are many, intertwined critical paths, so I'm not sure what it would mean.  Hide the units without edges?

Also, some of the critical paths are just a happenstance of timing.  If cargo randomly picks a different unit from the waiting set, the critical path may end up looking very different.

> Have you checked around to see if there's any sort of open formats for this sort of data?

I looked around a little, and I didn't see anything that would be usable.  If anyone has any ideas, I'd like to try them.  I also considered using one of the javascript graphing libraries, but most of them didn't fit, and I wanted to avoid heavy dependencies.  One thought I had was to add the JSON output and then allow other tools to process and display it.  I could also move the actual HTML part of this PR to a separate repo, but I figured its utility would be severely reduced.

> flatten this whole graph

My original version was exactly this (it had one row per cpu).  It had some drawbacks.  You can't visualize how the dependency edges work (particularly for pipelining).  It also ends up just being a solid block of boxes until the end when there is 1 or 2 left.  I had mouseover tooltips to help show more information, but it just didn't seem that useful. 

The alternative I intended was the graph at the bottom tells you whether or not it is maximizing the concurrency (via the green line).  But it is a bit difficult to visualize.  I also had a mouseover tooltip on the graph line to show which units were currently active, but I didn't finish it and took it out to get earlier feedback on whether this is generally useful.

I'm happy to continue discussing ideas for what to display and more interactive features.  I just wanted to see if this is something we'd like to add at all.  I'd like to focus on *actionable information* — that is what decisions and changes would a user make after reading this information.  Seeing the graphs can be "nifty", but if it doesn't help make changes, there isn't much value.  And the user does not have very many options (removing dependencies, splitting libraries into smaller pieces, removing features, etc.).

alexcrichton commented:
> I just wanted to see if this is something we'd like to add at all. I'd like to focus on actionable information — that is what decisions and changes would a user make after reading this information. 

I personally very much agree with this, and I should probably frame my comments moreso in how I think that this sort of information will be interpreted. When I personally analyze graphs like this I'm looking for a few things:

* Where is there idle CPU parallelism to be taken advantage of? For example is a dependency compiled very late in the dependency graph when it actually should be built much earlier? For example `im-rc` is built super later in Cargo's graph, but that's only because its build script depends on `rustc_version` which depends on `serde` which has its own big dependency tree. By removing the dependency on `rustc_version` we could shift most of this bottlenecking build time from the end of the build to idle CPU in the middle.

  This is what I was getting at with a more flat view where you might be able to more visually see what's happening for parallelism during the build. We ideally want "entirely full until the end" but most projects don't end up with that sort of graph right now.

* What dependencies take a particularly long time in rustc? These are good candidates for splitting up, trying to move more to monomorphization, etc. In general long-compiling crates should have as few dependencies as possible and scheduled as early in the build as possible. Or in other words, a build should just minimize the number of "big crates" it has.

* Why did it take so long to get my final crate to start compiling? What was the blocker to getting it running, aka what dependency has the chain that took the longest to compile. This is sort of what I was getting at with the critical path. I definitely agree that the critical path isn't the same between builds and is pretty nondeterministic, but I suspect many crate graphs have a pretty deterministic critical path, and raising awareness about that seems pretty useful. For example Cargo's critical path largely included cc-rs -> libgit2-sys -> git2 -> git2-curl -> cargo, and that's because libgit2-sys took so darned long to compile since it wasn't parallel.

I think that the current implementation you've got here is good enough for most of these metrics above. The reason I was curious about open source tools is that exploring the current graph is pretty difficult. For example my browser (Firefox, maybe it's buggy?) gets super janky when I'm scrolling around the giant SVG. Additionally it's difficult to see what unlocked a dependency (the line drawn into it) as well as what it unlocked because the units it connects to may be either above or below the fold. These aren't really things I expect us to fix, but it'd be cool if there was an open viewer for this that already solved these problems, but I wasn't quite able to find one!

FWIW I think it's also worth pointing out that the green line is an approximation of parallelism but it doesn't take into account internal parallelism in rustc (because it can't, really). That may not matter a huge amount for most projects though.

Overall I'd be fine merging this basically as-is with a few small improvements in the inline comments above. It's pretty low-risk in terms of maintenance and if we ever want to remove it it's not that hard to remove.

bors commented:
:umbrella: The latest upstream changes (presumably #7351) made this pull request unmergeable. Please resolve the merge conflicts.

ehuss commented on src/doc/src/reference/unstable.md:
```diff
@@ -255,3 +255,23 @@ etc.).
+The `-Ztimings` flag can optionally take a comma-separated list of the
+following values:
+
+- `html` — Saves a file called `cargo-timing.html` to the current directory
```
Filenames now include a timestamp.

ehuss commented on src/doc/src/reference/unstable.md:
```diff
@@ -255,3 +255,23 @@ etc.).
+- `html` — Saves a file called `cargo-timing.html` to the current directory
+  with a report of the compilation.
+- `info` — Displays a message to stdout after each compilation finishes with
+  how long it took.
```
This is mostly doing what rustbuild does (with RUSTC-TIMING).  Not sure if it can actually replace it, but if it was valuable there it might be useful for others?

alexcrichton replied:
Eh I think it's fine for now, it's always something we can iterate on

ehuss commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -0,0 +1,861 @@
+    /// A string describing the cargo target.
+    target: String,
+    /// The time when this unit started as an offset in seconds from `Timings::start`.
+    start: f64,
```
The end result needs float values to display and do the math, which requires more conversions, and makes it more cumbersome to repeatedly convert Duration to float.

ehuss commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -0,0 +1,861 @@
+        root_units: &[Unit<'_>],
+        profile: String,
+    ) -> Timings<'a, 'cfg> {
+        let mut root_map: HashMap<PackageId, Vec<String>> = HashMap::new();
```
I just put the whole thing behind an `Option`.

ehuss commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -0,0 +1,861 @@
+
+    /// Mark that the `.rmeta` file as generated.
+    pub fn unit_rmeta_finished(&mut self, id: u32, unlocked: Vec<&Unit<'a>>) {
+        if let Some(unit_time) = self.active.get_mut(&id) {
```
It cannot.  "fresh" units always emit `Message::Finish`, but the timing code only populates its internal "active" map for dirty units.  I've added a comment explaining.

alexcrichton replied:
Ah ok, failing that could this be restructured to avoid the indentation?

```rust
let unit_time = match self.active.get_mut(&id) {
    Some(unit_time) => time,
    None => return,
};
// ...
```

ehuss commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -0,0 +1,861 @@
+    pub fn unit_rmeta_finished(&mut self, id: u32, unlocked: Vec<&Unit<'a>>) {
+        if let Some(unit_time) = self.active.get_mut(&id) {
+            let t = d_as_f64(self.start.elapsed());
+            unit_time.rmeta_time = Some(t - unit_time.start);
```
The end result needs float values to display and do the math, which requires more conversions, and makes it more cumbersome to repeatedly convert to float.

ehuss commented:
I switched to using canvas which should be faster. Let me know if it has any problems.  It also allows interactive controls, and I added a few as examples.

alexcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -558,14 +581,21 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
         }
-        self.queue.finish(unit, &artifact);
+        let unlocked = self.queue.finish(unit, &artifact);
+        if let Some(t) = &mut self.timings {
```
Oh so I actually liked the way this was earlier where you didn't have to write `if let` at all the callsites, could the `Option` be internal to the `timings` so from a calling perspective it's a bit more lightweight?

alexcrichton commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -0,0 +1,861 @@
+
+    /// Mark that a unit has finished running.
+    pub fn unit_finished(&mut self, id: u32, unlocked: Vec<&Unit<'a>>) {
+        if let Some(mut unit_time) = self.active.remove(&id) {
```
(given the above comment, could this be refactored a similar way to early-return if it's missing from the map?

alexcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1123,7 +1076,8 @@ struct OutputOptions {
     fn new<'a>(cx: &Context<'a, '_>, unit: &Unit<'a>) -> OutputOptions {
-        let look_for_metadata_directive = cx.rmeta_required(unit);
+        let look_for_metadata_directive =
+            cx.rmeta_required(unit) || cx.bcx.config.cli_unstable().timings.is_some();
```
Could the calculation for `rmeta_required` take the clause for timing into account? I think that' remove the need to update `add_error_format_and_color` as well and at least at a cursory glance it seems like it'd be fine to change that definition

alexcrichton commented:
Also I can confirm the canvas is indeed speedy, and the touch to color build script executions a different colors was nice!

alexcrichton commented on src/cargo/core/compiler/timings.rs:
```diff
@@ -0,0 +1,641 @@
+    <td><label for="scale">Scale:</label></td>
+  </tr>
+  <tr>
+    <td><input type="range" min="0" max="30" value="0" id="min-unit-time"></td>
```
FWIW I found this a little unwieldy to work with because it was doing a 1s interval, but could this perhaps be more like a 100ms interval?

ehuss commented:
Updated with review comments. Thanks!

>  CPU usage of the host system

I'd love to see that.  I was reluctant to add a dependency for this, since I suspect getting it to compile and work on all host platforms will be tricky.

I also experimented with changing jobserver to report what it thinks the current concurrency is, which could be more useful with parallel-rustc.  That's something we could also consider for later.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 8be10f7e9b8fe9c2c1ec1c0ce8d3ffd3020bc18d has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 8be10f7e9b8fe9c2c1ec1c0ce8d3ffd3020bc18d -->
<!-- homu: {"type":"Approved","sha":"8be10f7e9b8fe9c2c1ec1c0ce8d3ffd3020bc18d","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit 8be10f7e9b8fe9c2c1ec1c0ce8d3ffd3020bc18d with merge d764fff2da618b24fbf506229509e3d183050c2e...
<!-- homu: {"type":"BuildStarted","head_sha":"8be10f7e9b8fe9c2c1ec1c0ce8d3ffd3020bc18d","merge_sha":"d764fff2da618b24fbf506229509e3d183050c2e"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=8388)
Approved by: alexcrichton
Pushing d764fff2da618b24fbf506229509e3d183050c2e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=8388"},"merge_sha":"d764fff2da618b24fbf506229509e3d183050c2e"} -->

csmoe commented:
@ehuss the js render seems not happy with a huge project(~550 deps in toal)
<img width="679" alt="image" src="https://user-images.githubusercontent.com/35686186/65368612-0c2c6b00-dc76-11e9-9aa4-4b0f60a0bd23.png">

est31 commented:
@csmoe this sounds similar to #7388 and #7399 . Can you check whether it works with #7397 applied?

csmoe commented:
@est31 thank you, rebuilding :)

luser commented:
FYI, I know I had mentioned this before but I have [an old branch](https://github.com/luser/cargo/tree/profile-build-steps) laying around where I hacked cargo to output profiling info in Chrome's tracing format:
https://github.com/luser/cargo/commit/a241a1fed649db82f4b9fd699e419bd83537a0a3

There are a zillion formats for profiling/tracing info like this unfortunately. The nice thing about supporting someone else's format is that you can use other tools to explore it like Chrome's `chrome://tracing`, Firefox's [profiler.firefox.com](https://profiler.firefox.com/), or [SpeedScope](https://www.speedscope.app/). Speedscope's docs [mention a whole pile of formats](https://github.com/jlfwong/speedscope#supported-file-formats) it supports. That list includes [flamescope](https://github.com/coolreader18/flamescope) which is a Rust crate that can output [Speedscope's preferred format](https://github.com/jlfwong/speedscope/wiki/Importing-from-custom-sources#speedscopes-file-format), so it ought to be possible to refactor that crate such that you could use it in cargo to generate that output.
